File: server/email-otp-boundary.ts
import {
  objectDecoder,
  optionalDecoder,
  transformDecoder,
  type Decoder,
} from '../lib/codec/core';
import {
  emailOtpAddressDecoder,
  emailOtpChallengeIdDecoder,
  emailOtpCodeDecoder,
  emailOtpDigestDecoder,
  emailOtpRateLimitKeyDecoder,
  emailOtpSaltDecoder,
  type EmailOtpAddress,
  type EmailOtpChallengeId,
  type EmailOtpCode,
  type EmailOtpDigest,
  type EmailOtpRateLimitKey,
  type EmailOtpSalt,
} from '../lib/domain/email-otp';
import {
  accountIdDecoder,
  type AccountId,
  type VaultContext,
} from '../lib/domain/identity';
import {
  createEmailOtpChallenge,
  decideEmailOtpIdentityResolution,
  emailOtpChallengeDecoder,
  emailOtpIdentityRecordDecoder,
  invalidateEmailOtpDelivery,
  resendEmailOtpChallenge,
  verifyEmailOtpChallenge,
  type EmailOtpChallenge,
  type EmailOtpIdentityRecord,
  type EmailOtpIdentityResolution,
  type EmailOtpPurpose,
  type PendingEmailOtpChallenge,
} from './core/email-otp';
import { termsConsentCommandDecoder } from './terms-consent/public';
import type {
  SignupAdmissionPort,
  SignupAdmissionReceipt,
} from './signup-admission/public';

export type EmailOtpClockPort = {
  nowEpochSeconds: () => unknown;
};

export type EmailOtpSecretPort = {
  createChallengeId: () => Promise<unknown>;
  createCode: () => Promise<unknown>;
  createSalt: () => Promise<unknown>;
};

export type EmailOtpHashInput = {
  readonly challengeId: EmailOtpChallengeId;
  readonly address: EmailOtpAddress;
  readonly code: EmailOtpCode;
  readonly salt: EmailOtpSalt;
};

export type EmailOtpHasherPort = {
  /** Production adapters must use a server-held pepper and timing-safe checks. */
  createDigest: (input: EmailOtpHashInput) => Promise<unknown>;
  matchesDigest: (
    input: EmailOtpHashInput & { readonly expectedDigest: EmailOtpDigest },
  ) => Promise<unknown>;
};

export type EmailOtpChallengeStore = {
  insertPending: (challenge: PendingEmailOtpChallenge) => Promise<void>;
  findById: (challengeId: EmailOtpChallengeId) => Promise<unknown>;
  /** Atomically replaces only the expected version. */
  compareAndSwap: (
    challengeId: EmailOtpChallengeId,
    expectedVersion: number,
    next: EmailOtpChallenge,
  ) => Promise<unknown>;
};

export type EmailOtpRateLimitKeys = {
  readonly address: EmailOtpRateLimitKey;
  readonly network: EmailOtpRateLimitKey;
  readonly account?: EmailOtpRateLimitKey;
};

const emailOtpRateLimitKeysDecoder: Decoder<EmailOtpRateLimitKeys> =
  transformDecoder(
    objectDecoder({
      address: emailOtpRateLimitKeyDecoder,
      network: emailOtpRateLimitKeyDecoder,
      account: optionalDecoder(emailOtpRateLimitKeyDecoder),
    }),
    (keys): EmailOtpRateLimitKeys => ({
      address: keys.address,
      network: keys.network,
      ...(keys.account === undefined ? {} : { account: keys.account }),
    }),
  );

export type EmailOtpRateLimitPort = {
  /** Production adapters must reserve address, network, and account atomically. */
  reserve: (
    keys: EmailOtpRateLimitKeys,
    nowEpochSeconds: number,
  ) => Promise<unknown>;
};

export type EmailOtpAbuseKeyInput = {
  readonly address: EmailOtpAddress;
  readonly accountId?: AccountId;
  /** Opaque request metadata interpreted only by the outer adapter. */
  readonly networkContext: unknown;
};

export type EmailOtpAbuseKeyPort = {
  /**
   * Production adapters derive non-reversible HMAC keys from the normalized
   * address, trusted account context, and request network context. Keys must
   * never be accepted from a request body.
   */
  deriveKeys: (input: EmailOtpAbuseKeyInput) => Promise<unknown>;
};

export type EmailOtpDeliveryInput = {
  readonly challengeId: EmailOtpChallengeId;
  readonly address: EmailOtpAddress;
  readonly code: EmailOtpCode;
  readonly expiresAtEpochSeconds: number;
};

export type EmailDeliveryPort = {
  sendOtp: (input: EmailOtpDeliveryInput) => Promise<void>;
};

export type EmailOtpIdentityDirectory = {
  findByAddress: (address: EmailOtpAddress) => Promise<unknown>;
  findAccountIdByVerifiedContact: (
    address: EmailOtpAddress,
  ) => Promise<unknown>;
};

export type EmailOtpStartIntent = 'sign-in' | 'link-current-account';

export type EmailOtpStartResult =
  | { readonly kind: 'accepted'; readonly challengeId: EmailOtpChallengeId }
  | { readonly kind: 'failed'; readonly error: 'authentication-unavailable' };

export async function startEmailOtp(input: {
  readonly address: unknown;
  readonly intent: EmailOtpStartIntent;
  readonly signupTermsConsent?: unknown;
  readonly vaultContext?: VaultContext;
  readonly networkContext: unknown;
  readonly clock: EmailOtpClockPort;
  readonly secrets: EmailOtpSecretPort;
  readonly hasher: EmailOtpHasherPort;
  readonly challenges: EmailOtpChallengeStore;
  readonly abuseKeys: EmailOtpAbuseKeyPort;
  readonly rateLimits: EmailOtpRateLimitPort;
  readonly delivery: EmailDeliveryPort;
}): Promise<EmailOtpStartResult> {
  let rawChallengeId: unknown;
  try {
    rawChallengeId = await input.secrets.createChallengeId();
  } catch {
    return authenticationUnavailable();
  }
  const challengeId = emailOtpChallengeIdDecoder.decode(rawChallengeId);
  if (!challengeId.ok) return authenticationUnavailable();

  const accepted = (): EmailOtpStartResult => ({
    kind: 'accepted',
    challengeId: challengeId.value,
  });
  try {
    const address = emailOtpAddressDecoder.decode(input.address);
    const signupTermsConsent =
      input.signupTermsConsent === undefined
        ? undefined
        : termsConsentCommandDecoder.decode(input.signupTermsConsent);
    const now = decodeEpoch(input.clock.nowEpochSeconds());
    const purpose = purposeFromStart(input.intent, input.vaultContext);
    if (
      !address.ok ||
      now === undefined ||
      !purpose ||
      (signupTermsConsent !== undefined && !signupTermsConsent.ok) ||
      (purpose.kind === 'link' && signupTermsConsent !== undefined)
    ) {
      return accepted();
    }
    const keys = emailOtpRateLimitKeysDecoder.decode(
      await input.abuseKeys.deriveKeys({
        address: address.value,
        ...(purpose.kind === 'link' ? { accountId: purpose.accountId } : {}),
        networkContext: input.networkContext,
      }),
    );
    if (!keys.ok) return accepted();
    if (purpose.kind === 'link' && keys.value.account === undefined) {
      return accepted();
    }
    const reserved = await input.rateLimits.reserve(keys.value, now);
    if (reserved !== true) return accepted();

    const [rawCode, rawSalt] = await Promise.all([
      input.secrets.createCode(),
      input.secrets.createSalt(),
    ]);
    const code = emailOtpCodeDecoder.decode(rawCode);
    const salt = emailOtpSaltDecoder.decode(rawSalt);
    if (!code.ok || !salt.ok) return accepted();
    const digest = emailOtpDigestDecoder.decode(
      await input.hasher.createDigest({
        challengeId: challengeId.value,
        address: address.value,
        code: code.value,
        salt: salt.value,
      }),
    );
    if (!digest.ok) return accepted();
    const created = createEmailOtpChallenge({
      challengeId: challengeId.value,
      address: address.value,
      digest: digest.value,
      salt: salt.value,
      purpose,
      ...(signupTermsConsent === undefined
        ? {}
        : { signupTermsConsent: signupTermsConsent.value }),
      nowEpochSeconds: now,
    });
    if (created.kind === 'rejected') return accepted();
    await input.challenges.insertPending(created.challenge);
    try {
      await input.delivery.sendOtp({
        challengeId: challengeId.value,
        address: address.value,
        code: code.value,
        expiresAtEpochSeconds: created.challenge.expiresAtEpochSeconds,
      });
    } catch {
      try {
        await invalidateAfterDeliveryFailure(
          input.challenges,
          created.challenge,
          now,
        );
      } catch {
        return accepted();
      }
    }
    return accepted();
  } catch {
    return accepted();
  }
}

export type EmailOtpCompletionResult =
  | {
      readonly kind: 'resolved';
      readonly resolution: Exclude<
        EmailOtpIdentityResolution,
        { readonly kind: 'rejected' | 'provision-account' }
      >;
    }
  | {
      readonly kind: 'admitted';
      readonly admission: SignupAdmissionReceipt;
    }
  | { readonly kind: 'failed'; readonly error: 'verification-failed' };

export async function completeEmailOtp(input: {
  readonly challengeId: unknown;
  readonly code: unknown;
  readonly clock: EmailOtpClockPort;
  readonly hasher: EmailOtpHasherPort;
  readonly challenges: EmailOtpChallengeStore;
  readonly identities: EmailOtpIdentityDirectory;
  readonly signupAdmission?: SignupAdmissionPort;
}): Promise<EmailOtpCompletionResult> {
  const challengeId = emailOtpChallengeIdDecoder.decode(input.challengeId);
  const code = emailOtpCodeDecoder.decode(input.code);
  const now = decodeEpoch(input.clock.nowEpochSeconds());
  if (!challengeId.ok || !code.ok || now === undefined) {
    return verificationFailed();
  }
  try {
    const challenge = emailOtpChallengeDecoder.decode(
      await input.challenges.findById(challengeId.value),
    );
    if (!challenge.ok || challenge.value.kind !== 'pending') {
      return verificationFailed();
    }
    const matches = await input.hasher.matchesDigest({
      challengeId: challenge.value.challengeId,
      address: challenge.value.address,
      code: code.value,
      salt: challenge.value.salt,
      expectedDigest: challenge.value.digest,
    });
    if (typeof matches !== 'boolean') return verificationFailed();
    const decision = verifyEmailOtpChallenge({
      challenge: challenge.value,
      digestMatches: matches,
      nowEpochSeconds: now,
    });
    if (decision.challenge !== undefined) {
      const swapped = await input.challenges.compareAndSwap(
        challenge.value.challengeId,
        challenge.value.version,
        decision.challenge,
      );
      if (swapped !== true) return verificationFailed();
    }
    if (decision.kind !== 'verified') return verificationFailed();

    const [rawIdentity, rawContactAccountId] = await Promise.all([
      input.identities.findByAddress(challenge.value.address),
      input.identities.findAccountIdByVerifiedContact(challenge.value.address),
    ]);
    const existingIdentity = optionalIdentity(rawIdentity);
    const verifiedContactAccountId = optionalAccountId(rawContactAccountId);
    if (!existingIdentity.ok || !verifiedContactAccountId.ok) {
      return verificationFailed();
    }
    const resolution = decideEmailOtpIdentityResolution({
      purpose: challenge.value.purpose,
      address: challenge.value.address,
      ...(existingIdentity.value === undefined
        ? {}
        : { existingIdentity: existingIdentity.value }),
      ...(verifiedContactAccountId.value === undefined
        ? {}
        : { verifiedContactAccountId: verifiedContactAccountId.value }),
    });
    if (resolution.kind === 'rejected') return verificationFailed();
    if (resolution.kind === 'provision-account') {
      if (
        challenge.value.signupTermsConsent === undefined ||
        input.signupAdmission === undefined
      ) {
        return verificationFailed();
      }
      const admission = await input.signupAdmission.admit({
        identity: { kind: 'email-otp', address: resolution.address },
        termsConsent: challenge.value.signupTermsConsent,
      });
      return admission.kind === 'admitted'
        ? { kind: 'admitted', admission: admission.receipt }
        : verificationFailed();
    }
    return { kind: 'resolved', resolution };
  } catch {
    return verificationFailed();
  }
}

export type EmailOtpResendResult = {
  readonly kind: 'accepted';
};

export async function resendEmailOtp(input: {
  readonly challengeId: unknown;
  readonly networkContext: unknown;
  readonly clock: EmailOtpClockPort;
  readonly secrets: EmailOtpSecretPort;
  readonly hasher: EmailOtpHasherPort;
  readonly challenges: EmailOtpChallengeStore;
  readonly abuseKeys: EmailOtpAbuseKeyPort;
  readonly rateLimits: EmailOtpRateLimitPort;
  readonly delivery: EmailDeliveryPort;
}): Promise<EmailOtpResendResult> {
  const accepted = { kind: 'accepted' } as const;
  const challengeId = emailOtpChallengeIdDecoder.decode(input.challengeId);
  const now = decodeEpoch(input.clock.nowEpochSeconds());
  if (!challengeId.ok || now === undefined) return accepted;
  try {
    const challenge = emailOtpChallengeDecoder.decode(
      await input.challenges.findById(challengeId.value),
    );
    if (!challenge.ok || challenge.value.kind !== 'pending') return accepted;
    const keys = emailOtpRateLimitKeysDecoder.decode(
      await input.abuseKeys.deriveKeys({
        address: challenge.value.address,
        ...(challenge.value.purpose.kind === 'link'
          ? { accountId: challenge.value.purpose.accountId }
          : {}),
        networkContext: input.networkContext,
      }),
    );
    if (!keys.ok) return accepted;
    if (
      challenge.value.purpose.kind === 'link' &&
      keys.value.account === undefined
    ) {
      return accepted;
    }
    const [rawCode, rawSalt] = await Promise.all([
      input.secrets.createCode(),
      input.secrets.createSalt(),
    ]);
    const code = emailOtpCodeDecoder.decode(rawCode);
    const salt = emailOtpSaltDecoder.decode(rawSalt);
    if (!code.ok || !salt.ok) return accepted;
    const digest = emailOtpDigestDecoder.decode(
      await input.hasher.createDigest({
        challengeId: challenge.value.challengeId,
        address: challenge.value.address,
        code: code.value,
        salt: salt.value,
      }),
    );
    if (!digest.ok) return accepted;
    const decision = resendEmailOtpChallenge({
      challenge: challenge.value,
      digest: digest.value,
      salt: salt.value,
      nowEpochSeconds: now,
    });
    if (decision.kind !== 'resent') {
      if (decision.challenge !== undefined) {
        await input.challenges.compareAndSwap(
          challenge.value.challengeId,
          challenge.value.version,
          decision.challenge,
        );
      }
      return accepted;
    }
    const reserved = await input.rateLimits.reserve(keys.value, now);
    if (reserved !== true) return accepted;
    const swapped = await input.challenges.compareAndSwap(
      challenge.value.challengeId,
      challenge.value.version,
      decision.challenge,
    );
    if (swapped !== true) return accepted;
    try {
      await input.delivery.sendOtp({
        challengeId: challenge.value.challengeId,
        address: challenge.value.address,
        code: code.value,
        expiresAtEpochSeconds: decision.challenge.expiresAtEpochSeconds,
      });
    } catch {
      try {
        await invalidateAfterDeliveryFailure(
          input.challenges,
          decision.challenge,
          now,
        );
      } catch {
        return accepted;
      }
    }
    return accepted;
  } catch {
    return accepted;
  }
}

type OptionalDecoded<T> =
  | { readonly ok: true; readonly value: T | undefined }
  | { readonly ok: false };

function optionalIdentity(
  input: unknown,
): OptionalDecoded<EmailOtpIdentityRecord> {
  if (input === undefined) return { ok: true, value: undefined };
  const decoded = emailOtpIdentityRecordDecoder.decode(input);
  return decoded.ok ? { ok: true, value: decoded.value } : { ok: false };
}

function optionalAccountId(input: unknown): OptionalDecoded<AccountId> {
  if (input === undefined) return { ok: true, value: undefined };
  const decoded = accountIdDecoder.decode(input);
  return decoded.ok ? { ok: true, value: decoded.value } : { ok: false };
}

function purposeFromStart(
  intent: EmailOtpStartIntent,
  context: VaultContext | undefined,
): EmailOtpPurpose | undefined {
  if (intent === 'sign-in') return { kind: 'sign-in' };
  return context ? { kind: 'link', accountId: context.accountId } : undefined;
}

async function invalidateAfterDeliveryFailure(
  challenges: EmailOtpChallengeStore,
  challenge: PendingEmailOtpChallenge,
  nowEpochSeconds: number,
): Promise<void> {
  const invalidated = invalidateEmailOtpDelivery(challenge, nowEpochSeconds);
  if (invalidated === undefined) return;
  await challenges.compareAndSwap(
    challenge.challengeId,
    challenge.version,
    invalidated,
  );
}

function decodeEpoch(input: unknown): number | undefined {
  return typeof input === 'number' && Number.isSafeInteger(input) && input >= 0
    ? input
    : undefined;
}

function authenticationUnavailable(): EmailOtpStartResult {
  return { kind: 'failed', error: 'authentication-unavailable' };
}

function verificationFailed(): EmailOtpCompletionResult {
  return { kind: 'failed', error: 'verification-failed' };
}

File: backend/internal/identity/email_otp_boundary.go
package identity

import "context"

type EmailOtpClock interface {
	NowEpochSeconds() int64
}

type EmailOtpSecretPort interface {
	CreateChallengeID(context.Context) (string, error)
	CreateCode(context.Context) (string, error)
	CreateSalt(context.Context) (string, error)
}

type EmailOtpHashInput struct {
	ChallengeID EmailOtpChallengeID
	Address     EmailOtpAddress
	Code        EmailOtpCode
	Salt        EmailOtpSalt
}

type EmailOtpHasherPort interface {
	CreateDigest(context.Context, EmailOtpHashInput) (string, error)
	MatchesDigest(context.Context, EmailOtpHashInput, EmailOtpDigest) (bool, error)
}

type EmailOtpChallengeStore interface {
	InsertPending(context.Context, EmailOtpChallenge) error
	FindByID(context.Context, EmailOtpChallengeID) (*EmailOtpChallenge, error)
	CompareAndSwap(context.Context, EmailOtpChallengeID, int64, EmailOtpChallenge) (bool, error)
}

type EmailOtpRateLimitKeys struct {
	Address EmailOtpRateLimitKey
	Network EmailOtpRateLimitKey
	Account *EmailOtpRateLimitKey
}

func (keys EmailOtpRateLimitKeys) Valid(requireAccount bool) bool {
	if _, err := ParseEmailOtpRateLimitKey(string(keys.Address)); err != nil {
		return false
	}
	if _, err := ParseEmailOtpRateLimitKey(string(keys.Network)); err != nil {
		return false
	}
	if keys.Account != nil {
		if _, err := ParseEmailOtpRateLimitKey(string(*keys.Account)); err != nil {
			return false
		}
	}
	return !requireAccount || keys.Account != nil
}

type EmailOtpAbuseKeyPort interface {
	DeriveKeys(context.Context, EmailOtpAddress, string, *AccountID) (EmailOtpRateLimitKeys, error)
}

type EmailOtpRateLimitPort interface {
	Reserve(context.Context, EmailOtpRateLimitKeys, int64) (bool, error)
}

type EmailOtpDeliveryInput struct {
	ChallengeID EmailOtpChallengeID
	Address     EmailOtpAddress
	Code        EmailOtpCode
	ExpiresAt   int64
}

type EmailOtpDeliveryPort interface {
	SendOtp(context.Context, EmailOtpDeliveryInput) error
}

type EmailOtpIdentityDirectory interface {
	FindByAddress(context.Context, EmailOtpAddress) (*EmailOtpIdentityRecord, error)
	FindAccountIDByVerifiedEmail(context.Context, VerifiedEmailAddress) (*AccountID, error)
}

type EmailOtpStartIntent string

const (
	EmailOtpStartSignIn             EmailOtpStartIntent = "sign-in"
	EmailOtpStartLinkCurrentAccount EmailOtpStartIntent = "link-current-account"
)

type EmailOtpStartResult struct {
	Accepted    bool
	ChallengeID EmailOtpChallengeID
	Error       string
}

func StartEmailOtp(ctx context.Context, input struct {
	Address             string
	Intent              EmailOtpStartIntent
	SignupTermsConsent  *SignupTermsConsent
	VaultContext        *VaultContext
	TrustedNetworkScope string
	Clock               EmailOtpClock
	Secrets             EmailOtpSecretPort
	Hasher              EmailOtpHasherPort
	Challenges          EmailOtpChallengeStore
	AbuseKeys           EmailOtpAbuseKeyPort
	RateLimits          EmailOtpRateLimitPort
	Delivery            EmailOtpDeliveryPort
}) EmailOtpStartResult {
	failure := EmailOtpStartResult{Error: "authentication-unavailable"}
	if input.Secrets == nil {
		return failure
	}
	rawChallengeID, err := input.Secrets.CreateChallengeID(ctx)
	if err != nil {
		return failure
	}
	challengeID, err := ParseEmailOtpChallengeID(rawChallengeID)
	if err != nil {
		return failure
	}
	accepted := EmailOtpStartResult{Accepted: true, ChallengeID: challengeID}
	if input.Clock == nil || input.Hasher == nil || input.Challenges == nil || input.AbuseKeys == nil ||
		input.RateLimits == nil || input.Delivery == nil || input.TrustedNetworkScope == "" {
		return accepted
	}
	address, err := ParseEmailOtpAddress(input.Address)
	if err != nil {
		return accepted
	}
	now := input.Clock.NowEpochSeconds()
	if !validTimestamp(now) {
		return accepted
	}
	purpose, accountID, ok := emailOtpPurposeFromStart(input.Intent, input.VaultContext)
	if !ok || (purpose.Kind == EmailOtpPurposeLink && input.SignupTermsConsent != nil) ||
		(input.SignupTermsConsent != nil && !input.SignupTermsConsent.Valid()) {
		return accepted
	}
	keys, err := input.AbuseKeys.DeriveKeys(ctx, address, input.TrustedNetworkScope, accountID)
	if err != nil || !keys.Valid(purpose.Kind == EmailOtpPurposeLink) {
		return accepted
	}
	reserved, err := input.RateLimits.Reserve(ctx, keys, now)
	if err != nil || !reserved {
		return accepted
	}
	rawCode, err := input.Secrets.CreateCode(ctx)
	if err != nil {
		return accepted
	}
	rawSalt, err := input.Secrets.CreateSalt(ctx)
	if err != nil {
		return accepted
	}
	code, err := ParseEmailOtpCode(rawCode)
	if err != nil {
		return accepted
	}
	salt, err := ParseEmailOtpSalt(rawSalt)
	if err != nil {
		return accepted
	}
	rawDigest, err := input.Hasher.CreateDigest(ctx, EmailOtpHashInput{
		ChallengeID: challengeID, Address: address, Code: code, Salt: salt,
	})
	if err != nil {
		return accepted
	}
	digest, err := ParseEmailOtpDigest(rawDigest)
	if err != nil {
		return accepted
	}
	decision := CreateEmailOtpChallenge(EmailOtpChallenge{
		ChallengeID: challengeID, Address: address, Digest: digest, Salt: salt,
		Purpose: purpose, SignupTermsConsent: input.SignupTermsConsent, CreatedAtEpochSeconds: now,
	})
	if !decision.Created {
		return accepted
	}
	if err := input.Challenges.InsertPending(ctx, decision.Challenge); err != nil {
		return accepted
	}
	if err := input.Delivery.SendOtp(ctx, EmailOtpDeliveryInput{
		ChallengeID: challengeID, Address: address, Code: code, ExpiresAt: decision.Challenge.ExpiresAtEpochSeconds,
	}); err != nil {
		if invalidated, valid := InvalidateEmailOtpDelivery(decision.Challenge, now); valid {
			_, _ = input.Challenges.CompareAndSwap(ctx, challengeID, decision.Challenge.Version, invalidated)
		}
	}
	return accepted
}

func emailOtpPurposeFromStart(intent EmailOtpStartIntent, vaultContext *VaultContext) (EmailOtpPurpose, *AccountID, bool) {
	switch intent {
	case EmailOtpStartSignIn:
		return EmailOtpPurpose{Kind: EmailOtpPurposeSignIn}, nil, true
	case EmailOtpStartLinkCurrentAccount:
		if vaultContext == nil || !validOidcVaultContext(*vaultContext) {
			return EmailOtpPurpose{}, nil, false
		}
		accountID := vaultContext.AccountID
		return EmailOtpPurpose{Kind: EmailOtpPurposeLink, AccountID: accountID}, &accountID, true
	default:
		return EmailOtpPurpose{}, nil, false
	}
}

type EmailOtpCompletionKind string

const (
	EmailOtpCompletionResolved EmailOtpCompletionKind = "resolved"
	EmailOtpCompletionAdmitted EmailOtpCompletionKind = "admitted"
	EmailOtpCompletionFailed   EmailOtpCompletionKind = "failed"
)

type EmailOtpCompletionResult struct {
	Kind       EmailOtpCompletionKind
	Error      string
	Resolution EmailOtpIdentityResolution
	Admission  SignupAdmissionReceipt
}

func CompleteEmailOtp(ctx context.Context, input struct {
	ChallengeID string
	Code        string
	Clock       EmailOtpClock
	Hasher      EmailOtpHasherPort
	Challenges  EmailOtpChallengeStore
	Identities  EmailOtpIdentityDirectory
	Signup      SignupAdmissionPort
}) EmailOtpCompletionResult {
	failure := EmailOtpCompletionResult{Kind: EmailOtpCompletionFailed, Error: "verification-failed"}
	if input.Clock == nil || input.Hasher == nil || input.Challenges == nil || input.Identities == nil {
		return failure
	}
	challengeID, err := ParseEmailOtpChallengeID(input.ChallengeID)
	if err != nil {
		return failure
	}
	code, err := ParseEmailOtpCode(input.Code)
	if err != nil {
		return failure
	}
	now := input.Clock.NowEpochSeconds()
	if !validTimestamp(now) {
		return failure
	}
	challenge, err := input.Challenges.FindByID(ctx, challengeID)
	if err != nil || challenge == nil || !ValidEmailOtpChallenge(*challenge) || challenge.Kind != EmailOtpPending {
		return failure
	}
	matches, err := input.Hasher.MatchesDigest(ctx, EmailOtpHashInput{
		ChallengeID: challenge.ChallengeID, Address: challenge.Address,
		Code: code, Salt: challenge.Salt,
	}, challenge.Digest)
	if err != nil {
		return failure
	}
	decision := VerifyEmailOtpChallenge(*challenge, matches, now)
	if decision.Changed {
		swapped, swapErr := input.Challenges.CompareAndSwap(ctx, challenge.ChallengeID, challenge.Version, decision.Challenge)
		if swapErr != nil || !swapped {
			return failure
		}
	}
	if !decision.Verified {
		return failure
	}
	existingIdentity, err := input.Identities.FindByAddress(ctx, challenge.Address)
	if err != nil || (existingIdentity != nil && !existingIdentity.Valid()) {
		return failure
	}
	verifiedEmailOwner, err := input.Identities.FindAccountIDByVerifiedEmail(ctx, challenge.Address.Verified())
	if err != nil || (verifiedEmailOwner != nil && !validAccountID(*verifiedEmailOwner)) {
		return failure
	}
	resolution := DecideEmailOtpIdentityResolution(struct {
		Purpose                     EmailOtpPurpose
		Address                     EmailOtpAddress
		ExistingIdentity            *EmailOtpIdentityRecord
		VerifiedEmailOwnerAccountID *AccountID
	}{
		Purpose: challenge.Purpose, Address: challenge.Address,
		ExistingIdentity: existingIdentity, VerifiedEmailOwnerAccountID: verifiedEmailOwner,
	})
	if resolution.Kind == EmailOtpIdentityRejected {
		return failure
	}
	if resolution.Kind == EmailOtpProvisionAccount {
		if challenge.SignupTermsConsent == nil || input.Signup == nil {
			return failure
		}
		verifiedIdentity := VerifiedSignupIdentity{
			Kind: SignupIdentityEmailOtp, Address: resolution.Address,
		}
		admission, admitErr := input.Signup.Admit(ctx, verifiedIdentity, *challenge.SignupTermsConsent)
		if admitErr != nil || !admission.Admitted ||
			!signupAdmissionMatchesRequest(admission.Receipt, verifiedIdentity, *challenge.SignupTermsConsent) {
			return failure
		}
		return EmailOtpCompletionResult{Kind: EmailOtpCompletionAdmitted, Admission: admission.Receipt}
	}
	return EmailOtpCompletionResult{Kind: EmailOtpCompletionResolved, Resolution: resolution}
}

type EmailOtpResendResult struct {
	Accepted bool
}

func ResendEmailOtp(ctx context.Context, input struct {
	ChallengeID         string
	TrustedNetworkScope string
	Clock               EmailOtpClock
	Secrets             EmailOtpSecretPort
	Hasher              EmailOtpHasherPort
	Challenges          EmailOtpChallengeStore
	AbuseKeys           EmailOtpAbuseKeyPort
	RateLimits          EmailOtpRateLimitPort
	Delivery            EmailOtpDeliveryPort
}) EmailOtpResendResult {
	accepted := EmailOtpResendResult{Accepted: true}
	if input.Clock == nil || input.Secrets == nil || input.Hasher == nil || input.Challenges == nil ||
		input.AbuseKeys == nil || input.RateLimits == nil || input.Delivery == nil || input.TrustedNetworkScope == "" {
		return accepted
	}
	challengeID, err := ParseEmailOtpChallengeID(input.ChallengeID)
	if err != nil {
		return accepted
	}
	now := input.Clock.NowEpochSeconds()
	if !validTimestamp(now) {
		return accepted
	}
	challenge, err := input.Challenges.FindByID(ctx, challengeID)
	if err != nil || challenge == nil || !ValidEmailOtpChallenge(*challenge) || challenge.Kind != EmailOtpPending {
		return accepted
	}
	var accountID *AccountID
	if challenge.Purpose.Kind == EmailOtpPurposeLink {
		value := challenge.Purpose.AccountID
		accountID = &value
	}
	keys, err := input.AbuseKeys.DeriveKeys(ctx, challenge.Address, input.TrustedNetworkScope, accountID)
	if err != nil || !keys.Valid(accountID != nil) {
		return accepted
	}
	reserved, err := input.RateLimits.Reserve(ctx, keys, now)
	if err != nil || !reserved {
		return accepted
	}
	rawCode, err := input.Secrets.CreateCode(ctx)
	if err != nil {
		return accepted
	}
	rawSalt, err := input.Secrets.CreateSalt(ctx)
	if err != nil {
		return accepted
	}
	code, err := ParseEmailOtpCode(rawCode)
	if err != nil {
		return accepted
	}
	salt, err := ParseEmailOtpSalt(rawSalt)
	if err != nil {
		return accepted
	}
	rawDigest, err := input.Hasher.CreateDigest(ctx, EmailOtpHashInput{
		ChallengeID: challenge.ChallengeID, Address: challenge.Address, Code: code, Salt: salt,
	})
	if err != nil {
		return accepted
	}
	digest, err := ParseEmailOtpDigest(rawDigest)
	if err != nil {
		return accepted
	}
	decision := ResendEmailOtpChallenge(*challenge, digest, salt, now)
	if decision.Changed {
		swapped, swapErr := input.Challenges.CompareAndSwap(ctx, challenge.ChallengeID, challenge.Version, decision.Challenge)
		if swapErr != nil || !swapped {
			return accepted
		}
	}
	if !decision.Resent {
		return accepted
	}
	if err := input.Delivery.SendOtp(ctx, EmailOtpDeliveryInput{
		ChallengeID: challenge.ChallengeID, Address: challenge.Address,
		Code: code, ExpiresAt: decision.Challenge.ExpiresAtEpochSeconds,
	}); err != nil {
		if invalidated, valid := InvalidateEmailOtpDelivery(decision.Challenge, now); valid {
			_, _ = input.Challenges.CompareAndSwap(ctx, challenge.ChallengeID, decision.Challenge.Version, invalidated)
		}
	}
	return accepted
}
